week3/homework/src/functions.js

```diff
@@ -0,0 +1,196 @@
+const fs = require('fs');
+const data = require('./index');
+const path = './to-dos.json';
+const fileName = 'to-dos.json';
+const uuidv4 = require('uuid/v4');
+const defaultFileContent = { todoArray: [] };
+
+function isFileExist() {
+  try {
+    if (fs.existsSync(path)) {
+      return true;
+    } else {
+      fs.writeFileSync(fileName, JSON.stringify(defaultFileContent));
+      return false;
+    }
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function isInputValid(request) {
+  try {
+    if (request.headers['content-type'] === 'application/json') {
+      if (request.body.hasOwnProperty('todo')) {
+        if (request.body.todo.hasOwnProperty('description')) {
+          if (
+            request.body.todo.description !== undefined &&
+            request.body.todo.description !== null &&
+            request.body.todo.description !== ''
+          ) {
+            return true;
+          }
+        }
+      }
+    } else return false;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function help() {
+  try {
+    return JSON.parse(fs.readFileSync('help.json', 'utf8'));
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function list() {
+  try {
+    if (isFileExist()) {
+      return JSON.parse(fs.readFileSync(fileName, 'utf8'));
+    }
+    return defaultFileContent;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function create(request) {
+  try {
+    if (isFileExist()) {
+      if (isInputValid(request)) {
+        request.body.todo.id = uuidv4();
+        request.body.todo.done = false;
+        const obj = JSON.parse(fs.readFileSync(fileName, 'utf8'));
+        obj.todoArray.push(request.body);
+        fs.writeFileSync(fileName, JSON.stringify(obj));
+        return obj;
+      }
+      return { error: 'content-type is not application/json or todo text is not valid' };
+    }
+    return defaultFileContent;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function getToDoById(id) {
+  try {
+    if (isFileExist()) {
+      const obj = JSON.parse(fs.readFileSync(fileName, 'utf8'));
+      const result = obj.todoArray.filter(element => {
+        if (element.todo.id == id) {
+          return element;
+        }
+      });
+      if (result.length == 0) {
+        return { error: 'There is no id=' + id + ' in todo list' };
+      } else {
+        return result[0];
+      }
+    }
+    return defaultFileContent;
+  } catch (error) {
+    return { error: `${error.message}` };
+  }
+}
+
+function markToDo(id, method) {
+  try {
+    if (isFileExist()) {
+      const obj = JSON.parse(fs.readFileSync(fileName, 'utf8'));
+      const result = obj.todoArray.filter(element => {
```

Comment:
why not use forEach ?

Reply:
Filter let you provide a callback for every element and returns a filtered array.
The main difference between forEach and filter is that forEach just loop over the array and executes the callback but filter executes the callback and check its return value. If the value is true element remains in the resulting array but if the return value is false the element will be removed for the resulting array..

After filtered the array I get an new array that provides some conditions . And then I am checking the if new array has an item or not.. If new array's length == 0, new array has no item.. I mean filtered array return empty array..